checker/cost.go

```diff
@@ -0,0 +1,430 @@
+	return CostEstimate{
+		multiplyByCostFactor(ce.Min, costPerUnit),
+		multiplyByCostFactor(ce.Max, costPerUnit),
+	}
+}
+
+// addUint64NoOverflow adds non-negative ints. If the result is exceeds math.MaxUint64, math.MaxUint64
+// is returned.
+func addUint64NoOverflow(x, y uint64) uint64 {
+	if y > 0 && x > math.MaxUint64-y {
+		return math.MaxUint64
+	}
+	return x + y
+}
+
+// multiplyUint64NoOverflow multiplies non-negative ints. If the result is exceeds math.MaxUint64, math.MaxUint64
+// is returned.
+func multiplyUint64NoOverflow(x, y uint64) uint64 {
+	if x > 0 && y > 0 && x > math.MaxUint64/y {
+		return math.MaxUint64
+	}
+	return x * y
+}
+
+// multiplyByFactor multiplies an integer by a cost factor float and returns the nearest integer value, rounded up.
+func multiplyByCostFactor(x uint64, y float64) uint64 {
+	xFloat := float64(x)
+	if xFloat > 0 && y > 0 && xFloat > math.MaxUint64/y {
+		return math.MaxUint64
+	}
+	return uint64(math.Ceil(xFloat * y))
+}
+
+var (
+	identCost  = CostEstimate{Min: 1, Max: 1}
+	selectCost = CostEstimate{Min: 1, Max: 1}
+	constCost  = CostEstimate{Min: 0, Max: 0}
+
+	createListBaseCost    = CostEstimate{Min: 10, Max: 10}
+	createMapBaseCost     = CostEstimate{Min: 30, Max: 30}
+	createMessageBaseCost = CostEstimate{Min: 40, Max: 40}
+)
+
+type coster struct {
+	env *Env
+	// map from Expr Id to field path. Only set for Exprs where the path from a root variable is known.
+	exprPath map[int64][]string
+	// map from comprehension iterVar Decls to the iterRange Expr Ids they iterate across.
+	iterVarMapping map[*exprpb.Decl]int64
+	checker        *exprpb.CheckedExpr
+	estimator      CostEstimator
+}
+
+// Cost estimates the cost of the parsed CEL expression.
+// The parsedExpr is expected to have passed type checking. If it has not been
+// type checked, the results are undefined.
+func Cost(checker *exprpb.CheckedExpr,
+	env *Env, estimator CostEstimator) CostEstimate {
+	c := coster{
+		env:            env,
+		checker:        checker,
+		estimator:      estimator,
+		exprPath:       map[int64][]string{},
+		iterVarMapping: map[*exprpb.Decl]int64{},
+	}
+	return c.cost(checker.GetExpr())
+}
+
+func (c *coster) cost(e *exprpb.Expr) CostEstimate {
+	if e == nil {
+		return CostEstimate{}
+	}
+	var cost CostEstimate
+	switch e.ExprKind.(type) {
+	case *exprpb.Expr_ConstExpr:
+		cost = constCost
+	case *exprpb.Expr_IdentExpr:
+		cost = c.costIdent(e)
+	case *exprpb.Expr_SelectExpr:
+		cost = c.costSelect(e)
+	case *exprpb.Expr_CallExpr:
+		cost = c.costCall(e)
+	case *exprpb.Expr_ListExpr:
+		cost = c.costCreateList(e)
+	case *exprpb.Expr_StructExpr:
+		cost = c.costCreateStruct(e)
+	case *exprpb.Expr_ComprehensionExpr:
+		cost = c.costComprehension(e)
+	default:
+		return CostEstimate{}
+	}
+	return cost
+}
+
+func (c *coster) costIdent(e *exprpb.Expr) CostEstimate {
+	identExpr := e.GetIdentExpr()
+
+	// resolve the field path
+	if ident := c.env.LookupIdent(identExpr.GetName()); ident != nil {
```

Comment:
The lookup here indicates whether something a variable or a type identifier, but by the time an expression is checked you can be guaranteed that the identifier will be found and that both have the same `identCost`.

Adding a reference to the `c.exprPath` for a type identifier won't do any harm, so I would remove the differentiation between variables and types with the `LookupIdent` call and just use the inner if-else block to differentiate between iter and non-iter identifiers.

Reply:
SGTM. I've tried to simplify this to match. Let me know if I mis-understood.